### PR TITLE
Fix `create-catalogs-schemas` to allow more than 1 level nesting more than the external location

### LINKED ISSUE #1700: [BUG]: create-catalogs-schemas fail if the location provided has more than 1 level nesting than the external location
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

I run `databricks labs ucx create-catalogs-schemas` and provide `s3://mohanatest/data/catalog/workspace_1961346165534048` as the managed storage location for a catalog. I have an external location for `s3://mohanatest/`.
The command fails with `ERROR [d.l.ucx.create-catalogs-schemas] NotFound: Failed to validate location for workspace_1961346165534048 catalog`.

### Expected Behavior

The catalog should get created as the existing external location works for the location provided.
But the current code only works for 1 level of nesting. Since I am providing 3 level of nesting sub-directories, it fails.
For example, location `s3://mohanatest/data` will work, but location `s3://mohanatest/data/catalog` fails. Ideally, both should succeed.

### Steps To Reproduce

1. Create external location for `s3://mohanatest/`
2. Run `databricks labs ucx create-catalogs-schemas` 
3. Enter `s3://mohanatest/data/catalog/workspace_1961346165534048` when prompted with "Please provide storage location url for catalog:workspace_1961346165534048."

### Cloud

AWS

### Operating System

macOS

### Version

latest via Databricks CLI

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by mohanab-db**
## Changes
Fix `create-catalogs-schemas` to allow more than 1 level nesting more than the external location.

### Linked issues

Resolves #1700

### Functionality 

- [ ] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`

### Tests

- [ ] manually tested
- [ ] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

File: tests/unit/hive_metastore/test_catalog_schema.py [hunk 1]
```diff
@@ -86,7 +94,7 @@ def prepare_test(ws, backend: MockBackend | None = None) -> CatalogSchema:
     return CatalogSchema(ws, table_mapping, principal_acl, backend)
 
 
-@pytest.mark.parametrize("location", ["s3://foo/bar", "s3://foo/bar/test"])
+@pytest.mark.parametrize("location", ["s3://foo/bar", "s3://foo/bar/test", "s3://foo/bar/test/1"])
 def test_create_all_catalogs_schemas_creates_catalogs(location: str):
     """Catalog 2 and 3 should be created; catalog 1 already exists."""
```
Comment (JCZuurmond) [inline tests/unit/hive_metastore/test_catalog_schema.py:99]:
```suggestion
    """Catalog 2-4 should be created; catalog 1 already exists."""
```

File: tests/unit/hive_metastore/test_catalog_schema.py [hunk 2]
```diff
@@ -70,6 +70,14 @@ def prepare_test(ws, backend: MockBackend | None = None) -> CatalogSchema:
                     'src_table': 'abfss://container@msft/path/dest3',
                     'workspace_name': 'workspace',
                 },
+                {
+                    'catalog_name': 'catalog4',
+                    'dst_schema': 'schema4',
+                    'dst_table': 'table1',
+                    'src_schema': 'schema1',
+                    'src_table': 'abfss://container@msft/path/dest3',
```
Comment (JCZuurmond) [inline tests/unit/hive_metastore/test_catalog_schema.py:78]:
```suggestion
                    'src_table': 'abfss://container@msft/path/dest4',
```

File: tests/unit/hive_metastore/test_catalog_schema.py [hunk 3]
```diff
@@ -86,7 +94,7 @@ def prepare_test(ws, backend: MockBackend | None = None) -> CatalogSchema:
     return CatalogSchema(ws, table_mapping, principal_acl, backend)
 
 
-@pytest.mark.parametrize("location", ["s3://foo/bar", "s3://foo/bar/test"])
+@pytest.mark.parametrize("location", ["s3://foo/bar", "s3://foo/bar/test", "s3://foo/bar/test/1"])
```
Comment (JCZuurmond) [inline tests/unit/hive_metastore/test_catalog_schema.py:97]: To show it is not about the integer
```suggestion
@pytest.mark.parametrize("location", ["s3://foo/bar", "s3://foo/bar/test", "s3://foo/bar/test/baz"])
```

File: src/databricks/labs/ucx/hive_metastore/catalog_schema.py [hunk 4]
```diff
@@ -144,8 +144,10 @@ def _validate_location(self, location: str):
         for external_location in self._external_locations:
             if location == external_location.url:
                 return True
-            if location_path.match(f"{external_location.url}/*"):
-                return True
+            if external_location.url is not None:
+                external_location_path = PurePath(external_location.url)
+                if location_path.parts[: len(external_location_path.parts)] == external_location_path.parts:
+                    return True
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/hive_metastore/catalog_schema.py:150]: Does this work?
```suggestion
            if location_path.match(f"{external_location.url}/**"):
                return True
```

**Review by JCZuurmond — COMMENTED** [review]
Hi @mohanab-db, thank you for your contribution again. I would like to see if we can stick with the glob pattern

File: src/databricks/labs/ucx/hive_metastore/catalog_schema.py (hunk 4 above)
Comment (mohanab-db) [reply src/databricks/labs/ucx/hive_metastore/catalog_schema.py:150]: @JCZuurmond This doesn't seem to work. However, this works `location_path.match(f"{external_location}/*/*")`. Again, this fails if another additional level of nesting is provided.

File: tests/unit/hive_metastore/test_catalog_schema.py (hunk 1 above)
Comment (mohanab-db) [reply tests/unit/hive_metastore/test_catalog_schema.py]: @JCZuurmond Thanks for reviewing. Updated this.

File: tests/unit/hive_metastore/test_catalog_schema.py (hunk 2 above)
Comment (mohanab-db) [reply tests/unit/hive_metastore/test_catalog_schema.py]: @JCZuurmond Thanks for reviewing. Updated this.

File: src/databricks/labs/ucx/hive_metastore/catalog_schema.py (hunk 4 above)
Comment (JCZuurmond) [reply src/databricks/labs/ucx/hive_metastore/catalog_schema.py:150]: hmmm, too bad, could you add a test for a deeply nested structure does not work? Like 8 levels deep

Comment (JCZuurmond) [reply src/databricks/labs/ucx/hive_metastore/catalog_schema.py:150]: `location_path.match(f"{external_location}/**/*")` What does that do? I would expect it to be solvable with glob pattern

Comment (mohanab-db) [reply src/databricks/labs/ucx/hive_metastore/catalog_schema.py]: @JCZuurmond Looks like ** is not supported by PurePath.match. However, I am able to get a match using fnmatch. `fnmatch.fnmatch(location, external_location*)`. Please review.

**Review by nkvuong — APPROVED** [review]

**Review by JCZuurmond — APPROVED** [review]

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**